src/abstract-interpretation/ainode.ts

```diff
@@ -0,0 +1,48 @@
+import { NodeId, ParentInformation, RNodeWithParent } from '../r-bridge'
+import { assertUnreachable, guard } from '../util/assert'
+import { Domain, unifyDomains } from './domain'
+
+export interface AINode {
+	// The ID of the node that logically holds the domain
+	readonly nodeId:       NodeId
+	// The ID of the whole expression that the domain was calculated from
+	readonly expressionId: NodeId
+	readonly domain:       Domain
+	readonly astNode:      RNodeWithParent<ParentInformation>
+}
+
+
+export class AINodeStore extends Map<NodeId, AINode> {
+	constructor(content: AINode[] | AINode | undefined = undefined) {
+		if(Array.isArray(content)) {
+			super(content.map(node => [node.nodeId, node]))
+		} else if(content !== undefined) {
+			super([[content.nodeId, content]])
+		} else if(content === undefined) {
+			super()
+		} else {
+			assertUnreachable(content)
+		}
+	}
+
+	push(node: AINode): void {
```

Comment:
given the map nature, something like `register` is probably a better name.

Reply:
\+ given that we add a wrapper to use, how about guarding against overwriting?
I mean yes, we assume that every node is unique, but what if we add the same node multiple times.

Reply:
Sure, fixed it :) I added a `guard(...)` that guards against attempts that want to overwrite a value. When I spot a case that triggers the guard, I'll reevaluate if this is correct behavior (but it's probably not).

Reply:
Nvm, we do indeed use `register` to overwrite values. I added an extra flag that should be set when overwriting should be possible.